src/plan.rs

```diff
@@ -0,0 +1,1235 @@
+// Miniscript
+// Written in 2022 by rust-miniscript developers
+//
+// To the extent possible under law, the author(s) have dedicated all
+// copyright and related and neighboring rights to this software to
+// the public domain worldwide. This software is distributed without
+// any warranty.
+//
+// You should have received a copy of the CC0 Public Domain Dedication
+// along with this software.
+// If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+//! A spending plan or *plan* for short is a representation of a particular spending path on a
+//! descriptor. This allows us to analayze a choice of spending path without producing any
+//! signatures or other witness data for it.
+//!
+//! To make a plan you provide the descriptor with "assets" like which keys you are able to use, hash
+//! pre-images you have access to, absolute/relative timelock constraints etc.
+//!
+//! Once you've got a plan it can tell you its expected satisfaction weight which can be useful for
+//! doing coin selection. Furthermore it provides which subset of those keys and hash pre-images you
+//! will actually need as well as what locktime or sequence number you need to set.
+//!
+//! Once you've obstained signatures, hash pre-images etc required by the plan, it can create a
+//! witness/script_sig for the input.
+
+use core::cmp::Ordering;
+use core::iter::FromIterator;
+
+use bitcoin::absolute::LockTime;
+use bitcoin::address::WitnessVersion;
+use bitcoin::hashes::{hash160, ripemd160, sha256};
+use bitcoin::key::XOnlyPublicKey;
+use bitcoin::script::PushBytesBuf;
+use bitcoin::taproot::{ControlBlock, LeafVersion, TapLeafHash};
+use bitcoin::{bip32, psbt, ScriptBuf, Sequence};
+
+use crate::descriptor::{self, Descriptor, DescriptorType, KeyMap};
+use crate::miniscript::hash256;
+use crate::miniscript::satisfy::{Placeholder, Satisfier, SchnorrSigType};
+use crate::prelude::*;
+use crate::util::witness_size;
+use crate::{DefiniteDescriptorKey, DescriptorPublicKey, Error, MiniscriptKey, ToPublicKey};
+
+/// Trait describing a present/missing lookup table for constructing witness templates
+///
+/// This trait mirrors the [`Satisfier`] trait with the difference that instead of returning the
+/// item if it's present, it only returns a boolean to indicate its presence.
```

Comment:
Not quite right anymore.

Reply:
Solved in 14833b0a07b01b7289237e0f35e411b0af65342f in #592